website/source/docs/internals/providers/resource-data.html.md

```diff
@@ -0,0 +1,180 @@
+---
+layout: "docs"
+page_title: "Using ResourceData"
+sidebar_current: "docs-internals-provider-guide-resource-data"
+description: |-
+  Details about the ubiquitous ResourceData type and all its options.
+---
+
+# Using `ResourceData`
+
+When working with the `Create`, `Read`, `Update`, `Destroy`, and `Exists`
+methods on your Provider, it's almost impossible to not run into
+`ResourceData`. It's a type that is used all over, so it's important to
+understand it. But `ResourceData` can also be tricky, because it's not
+abstracting a single logical concept in Terraform.
+
+## How to Conceptualise `ResourceData`
+
+The most useful way to think of `ResourceData` is not as a placeholder for any
+one concept in Terraform—it's not just information about the config or the
+state, for example—but as the desired state of an object. If you think of
+Terraform as the convergence of your config file and your infrastructure,
+`ResourceData` is how you express  that understanding of what your
+infrastructure should look like.
+
+## What `ResourceData` Abstracts
+
+To get practical, `ResourceData` abstracts four separate concepts into a single
+source of truth:
+
+* The current state
+* The config file
+* The diff being applied
+* Any calls to `ResourceData.Set`
+
+Those are expressed in the order of their priority. That is, the config
+overrides any values set in the state, and every time you call `Set`, it
+overrides everything else.
+
+`ResourceData` often gets confused for either the state or the config, but it's
+important to realise that it's a step removed from these concepts, and is used
+more as an understanding of what the infrastructure _should_ be.
+
+## Retrieving Properties
+
+To retrieve properties from `ResourceData`, use the `Get` or `GetOk` methods.
```

Comment:
Again, I see lots of text, but what is `GetOk`? How do I use it? Adding a 2-3 line code sample here would do wonders for the usefulness of this guide.